File: js/BeginnersSelection/08_Otoshidama.js
// https://atcoder.jp/contests/abs/tasks/abc085_c

Object.prototype.let = function(f) {
  return f(this.valueOf());
}

require('fs')
  .readFileSync('/dev/stdin', 'utf8')
  .split(' ')
  .map(x => +x)
  .let(([n, y]) => [n, y / 1000])
  .let(([n, y]) =>
    ((v, f) => v != null ? f(v) : v).let(oc => oc(
      [...Array(n + 1).keys()].find(a =>
        (y - n - 9 * a)
          .let(x => x % 4 == 0 && x >= 0 && n - a - x / 4 >= 0)
      ),
      a => ((y - n - 9 * a) / 4)
        .let(b => `${a} ${b} ${n - a - b}`)
    ))
    || '-1 -1 -1'
  )
  .let(console.log);

// Optional chaining is available in Node.js ver 14.

// https://atcoder.jp/contests/abs/submissions/21477540

File: jl/BeginnersSelection/08_Otoshidama.jl
# https://atcoder.jp/contests/abs/tasks/abc085_c

using Printf

(
  readline()
  |> split
  |> x -> parse.(Int, x)
  |> ((n, y),) -> (n, y / 1000)
  |> ((n, y),) -> (
    (0 : n)
    |> as -> findfirst(a -> (
      (y - n - 9 * a)
      |> x -> x % 4 == 0 && x >= 0 && n - a - x / 4 >= 0
    ), as)
    |> a -> (
      a != nothing
        ? (
          (a - 1)
          |> a ->
            ((y - n - 9 * a) / 4)
            |> Int
            |> b -> (a, b, n - a - b)
        )
        : (-1, -1, -1)
    )
  )
  |> (((a, b, c),) -> @sprintf("%d %d %d", a, b, c))
  |> println
)

# https://atcoder.jp/contests/abs/submissions/22649190
